design-documents/module-split.md

```diff
@@ -0,0 +1,83 @@
+    * CatalogApi
+    
+    The following is proposed naming for modules:
+    Admin Ui and admin related business logic (blocks, controllers, static resources, models and resources potentially)
+    * CatalogAdmin
+    * CatalogAdminUi
+    * CatalogBackoffice
+    
+    Storefront Ui and storefront related business logic (blocks, controllers, static resources, models and resources potentially)
+    * CatalogStorefrontUi
+    * CatalogStorefront
+    * CatalogStore
+    
+    Cron (models and resources potentially)
+    * CatalogCron
+    
+    CLI (models and resources potentially)
+    * CatalogCli
+    
+    WebApi (models and resources potentially)
+    * CatalogWebApi
+    
+    Common code (business logic used by multiple areas, models, resource modelss)
+    * CatalogCommon
+    * CatalogShared
+    * CatalogDomain
+    
+    Common static resources
+    * CatalogCommonUi
+
+    Recommended naming is the following: CatalogApi, CatalogAdminUi, CatalogStorefrontUi, CatalogCron, CatalogCli, CatalogCommon, CatalogCommonUi.
+
+2. Keep old or change namespaces
+
+    Per [service isolation proposal](service-isolation.md) each module will be split on the admin, storefront, cron, webapi, cli and common parts. For example, `Catalog`, will be split on `CatalogAdminUi`, `CatalogStorefrontUi`, `CatalogCron`, etc.
+    
+    For dealing with namespaces, there are the following options.
+    
+    1. Change, introduce new namespaces and preserve old ones for backwards compatibility. Fix namespaces in the modules. Old module Catalog will contain classes under old namespace, marked with @deprecated annotation, that extend classes with new namespaces and di configuration for backwards compatibility. Then in the future minor release these classes will and di configuration will be removed.
+        Pros
+        * We going to have correct namespaces after module split, for instance \Magento\CatalogAdminUi\Block\…
+        * Allows transition period
+        
+        Cons
+        * Introduce configuration for virtual types under new namespace that will duplicate configuration for virtual types with old namespace
+        * Duplicate configuration for types (params definition)
+        * Double configuration for autoloader (not sure if it’s concern, but would be good to measure)
+        * Amount of modules will double, will be harder to navigate codebase and more overhead to support until we deprecate modules with old namespaces
+        * May need to add new tests to make sure everything is consistent (edited) 
+        * One minor release probably is not enough time for developers to fix their extensions
+    
+    2. Change, introduce new namespaces and don't preserve backwards compatibility
+        Pros
+        * We going to have correct namespaces after module split
+        * No overhead from preserving keeping backwards compatibility
+        
+        Cons
+        * Backwards incompatible, all of the extensions will be affected
```

Comment:
This should not even be an option. You're talking about completely breaking every extension in the ecosystem overnight; causing every single extn developer to go fork their extn into two versions, one for older magento and one for this; and causing every develop to have to learn a new paradigm.  You know the costs of all that across the ecosystem is easily hundreds of thousands of dollars and right?